File: src/data/ble/decoder/lib/device.condition.ts
import { Logger } from '@nestjs/common';
import { Conditions, DecodeDevice } from './types';
import {
  And,
  Index,
  Inverse,
  MacAtIndex,
  ManufacturerData,
  Name,
  NoManufacturerData,
  Or,
  ReverseMacAtIndex,
  ServiceData,
  UUID,
} from './decoder.constants';

const logger: Logger = new Logger('deviceMatches');

export const deviceMatches = (
  device: DecodeDevice,
  conditions: Conditions,
  operand?: any,
): boolean => {
  if (conditions.length === 0) {
    return true;
  }
  if (conditions.includes(Or)) {
    return conditions
      .reduce(
        (acc, cond) => {
          if (cond === Or) {
            acc.push([]);
          } else {
            acc[acc.length - 1].push(cond);
          }
          return acc;
        },
        [[]] as Conditions[],
      )
      .some((conditionGroup) => deviceMatches(device, conditionGroup));
  }
  // logger.debug({
  //   device,
  //   condition: conditions[0],
  //   condition1: conditions[1],
  //   operand,
  // });
  switch (conditions[0]) {
    case ServiceData:
      return deviceMatches(device, conditions.slice(1), device.serviceData);
    case ManufacturerData:
      return deviceMatches(
        device,
        conditions.slice(1),
        device.manufacturerData,
      );
    case NoManufacturerData:
      return (
        device.manufacturerData.length === 0 &&
        deviceMatches(device, conditions.slice(1))
      );
    case Name:
      return deviceMatches(
        device,
        conditions.slice(1),
        (operand ?? device).name,
      );
    case UUID:
      return deviceMatches(
        device,
        conditions.slice(1),
        (operand ?? device).uuid,
      );
    case MacAtIndex:
      return deviceMatches(
        device,
        conditions.slice(2),
        device.macAddress[conditions[1]],
      );
    case ReverseMacAtIndex:
      return deviceMatches(
        device,
        conditions.slice(2),
        device.macAddress.split('').reverse().join('')[conditions[1]],
      );
    case And:
      return deviceMatches(device, conditions.slice(1));
    case Or:
      return true;
    case Inverse:
      return !deviceMatches(device, conditions.slice(1), operand);
    case Index:
      return deviceMatches(
        device,
        conditions.slice(2),
        operand.slice(
          typeof conditions[1] === 'string'
            ? Number.parseInt(conditions[1], 10)
            : conditions[1],
        ),
      );
    case '>=':
      return (
        (typeof operand === 'number' ? operand : operand.length) >=
          (conditions[1] as number) &&
        deviceMatches(device, conditions.slice(2), operand)
      );
    case '<=':
      return (
        (typeof operand === 'number' ? operand : operand.length) <=
          (conditions[1] as number) &&
        deviceMatches(device, conditions.slice(2), operand)
      );
    case '>':
      return (
        (typeof operand === 'number' ? operand : operand.length) >
          (conditions[1] as number) &&
        deviceMatches(device, conditions.slice(2), operand)
      );
    case '<':
      return (
        (typeof operand === 'number' ? operand : operand.length) <
          (conditions[1] as number) &&
        deviceMatches(device, conditions.slice(2), operand)
      );
    case '=':
      if (typeof conditions[1] === 'string') {
        return (
          operand.toString().startsWith(conditions[1]) &&
          deviceMatches(device, conditions.slice(2), operand)
        );
      }
      if (typeof operand === 'string') {
        return (
          operand.length === conditions[1] &&
          deviceMatches(device, conditions.slice(2), operand)
        );
      }
      return (
        conditions[1] === operand &&
        deviceMatches(device, conditions.slice(2), operand)
      );
    default:
      if (typeof conditions[0] === 'string') {
        return (
          operand.toString().startsWith(conditions[0]) &&
          deviceMatches(device, conditions.slice(1), operand)
        );
      }
      if (typeof operand === 'string') {
        return (
          operand.length === conditions[0] &&
          deviceMatches(device, conditions.slice(1), operand)
        );
      }
      return (
        conditions[0] === operand &&
        deviceMatches(device, conditions.slice(1), operand)
      );
      return false;
  }
};

File: govee-python/src/govee/data/ble/device_condition.py
"""Port of the TypeScript device.condition logic used by the BLE decoder.

This module implements device_matches(device, conditions) to evaluate
complex condition arrays against a decoded device. It mirrors the
behavior of lib/data/ble/decoder/lib/device.condition.ts used by the
TypeScript decoder so the Python decoder can reach feature parity.
"""
from __future__ import annotations

from typing import Any, List

And = "&"
Or = "|"
ServiceData = "servicedata"
ManufacturerData = "manufacturerdata"
NoManufacturerData = "no-mfgdata"
Name = "name"
UUID = "uuid"
MacAtIndex = "mac@index"
ReverseMacAtIndex = "revmac@index"
Index = "index"
Inverse = "inverse"


def device_matches(device: dict, conditions: List[Any], operand: Any = None) -> bool:
    if not conditions:
        return True

    if Or in conditions:
        groups: List[List[Any]] = [[]]
        for cond in conditions:
            if cond == Or:
                groups.append([])
            else:
                groups[-1].append(cond)
        return any(device_matches(device, g) for g in groups)

    head = conditions[0]
    # Service/Manufacturer data handling
    if head == ServiceData:
        return device_matches(device, conditions[1:], device.get('serviceData'))
    if head == ManufacturerData:
        return device_matches(device, conditions[1:], device.get('manufacturerData'))
    if head == NoManufacturerData:
        md = device.get('manufacturerData') or ''
        return (len(md) == 0) and device_matches(device, conditions[1:])
    if head == Name:
        return device_matches(device, conditions[1:], (operand or device).get('name'))
    if head == UUID:
        return device_matches(device, conditions[1:], (operand or device).get('uuid'))
    if head == MacAtIndex:
        mac = device.get('macAddress', '')
        idx = conditions[1]
        try:
            return device_matches(device, conditions[2:], mac[idx])
        except Exception:
            return False
    if head == ReverseMacAtIndex:
        mac = device.get('macAddress', '')
        try:
            rev = ''.join(reversed(list(mac)))
            return device_matches(device, conditions[2:], rev[conditions[1]])
        except Exception:
            return False
    if head == And:
        return device_matches(device, conditions[1:])
    if head == Or:
        return True
    if head == Inverse:
        return not device_matches(device, conditions[1:], operand)
    if head == Index:
        if operand is None:
            return False
        start = int(conditions[1]) if isinstance(conditions[1], str) else conditions[1]
        try:
            return device_matches(device, conditions[2:], operand[start:])
        except Exception:
            return False

    # length / comparison helpers
    if head == '>=':
        op_len = operand if isinstance(operand, (int, float)) else len(operand or '')
        return (op_len >= int(conditions[1])) and device_matches(device, conditions[2:], operand)
    if head == '<=':
        op_len = operand if isinstance(operand, (int, float)) else len(operand or '')
        return (op_len <= int(conditions[1])) and device_matches(device, conditions[2:], operand)
    if head == '>':
        op_len = operand if isinstance(operand, (int, float)) else len(operand or '')
        return (op_len > int(conditions[1])) and device_matches(device, conditions[2:], operand)
    if head == '<':
        op_len = operand if isinstance(operand, (int, float)) else len(operand or '')
        return (op_len < int(conditions[1])) and device_matches(device, conditions[2:], operand)

    if head == '=':
        # string prefix match
        if isinstance(conditions[1], str):
            try:
                return str(operand).startswith(conditions[1]) and device_matches(device, conditions[2:], operand)
            except Exception:
                return False
        if isinstance(operand, str):
            return (len(operand) == conditions[1]) and device_matches(device, conditions[2:], operand)
        return (conditions[1] == operand) and device_matches(device, conditions[2:], operand)

    # default: check literal equality or string-prefix when condition is a string
    if isinstance(head, str):
        try:
            return str(operand).startswith(head) and device_matches(device, conditions[1:], operand)
        except Exception:
            return False
    if isinstance(operand, str):
        return (len(operand) == head) and device_matches(device, conditions[1:], operand)

    return (head == operand) and device_matches(device, conditions[1:], operand)


__all__ = ["device_matches"]
